dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp

```diff
@@ -615,55 +544,41 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
         join_output_columns.emplace_back(match_helper_name, Join::match_helper_type);
     }
 
-    DataTypes join_key_types;
-    getJoinKeyTypes(join, join_key_types);
-    TiDB::TiDBCollators collators;
-    size_t join_key_size = join_key_types.size();
-    if (join.probe_types_size() == static_cast<int>(join_key_size) && join.build_types_size() == join.probe_types_size())
-        for (size_t i = 0; i < join_key_size; i++)
-        {
-            if (removeNullable(join_key_types[i])->isString())
-            {
-                if (join.probe_types(i).collate() != join.build_types(i).collate())
-                    throw TiFlashException("Join with different collators on the join key", Errors::Coprocessor::BadRequest);
-                collators.push_back(getCollatorFromFieldType(join.probe_types(i)));
-            }
-            else
-                collators.push_back(nullptr);
-        }
+    DataTypes join_key_types = JoinInterpreterHelper::getJoinKeyTypes(join);
+    TiDB::TiDBCollators collators = JoinInterpreterHelper::getJoinKeyCollators(join, join_key_types);
 
     Names left_key_names, right_key_names;
     String left_filter_column_name, right_filter_column_name;
 
-    /// add necessary transformation if the join key is an expression
+    /// generate necessary expression actions if the join key is an expression
 
-    prepareJoin(
+    auto probe_side_prepare_actions = prepareJoin(
         swap_join_side ? join.right_join_keys() : join.left_join_keys(),
         join_key_types,
-        left_pipeline,
+        probe_pipeline,
         left_key_names,
         true,
         is_tiflash_right_join,
         swap_join_side ? join.right_conditions() : join.left_conditions(),
         left_filter_column_name);
 
-    prepareJoin(
+    auto build_side_prepare_actions = prepareJoin(
         swap_join_side ? join.left_join_keys() : join.right_join_keys(),
         join_key_types,
-        right_pipeline,
+        build_pipeline,
         right_key_names,
```

Comment:
will you replace all left/right by probe/build? mixed using two kinds of names will confuse the reader.

Reply:
no.